### PR TITLE
Keep/use FOR_ALL_WINDOWS macro (which declares wp)

### PR DESCRIPTION
**Pull request by blueyed**
Ref: https://github.com/neovim/neovim/pull/6252/files#r303023078

### CODE REVIEW

**Comment by blueyed** [comment]
Is this good?  (waiting for feedback before fixing linting)

**Comment by blueyed** [comment]
I think this could be even upstreamed to Vim then (the declaration in the macro(s)).

**Comment by jamessan** [comment]
> I think this could be even upstreamed to Vim then (the declaration in the macro(s)).

It can't, because they don't require C99.

**Comment by justinmk** [comment]
I think the explicitness was good. Why spend time on this?

**Comment by blueyed** [comment]
> It can't, because they don't require C99.

Is that required for the declaration in the macro?

> I think the explicitness was good. Why spend time on this?

Using gcc9 produces a warning, but I could not reproduce this: I see a warning for `tautological-compare` when doing it explicitly, but not through the macro - but it happens within the macro on Travis with gcc9 from Ubuntu's ppa.
Might be related to https://github.com/gcc-mirror/gcc/commit/98de922f4, but I expect this to be in both my (Arch's) gcc 9.1 and Ubuntu's.

That's when I've noticed that there's quite a diff in this region in globals.h, and thought it would be nice to align this with Vim a bit.

**Comment by bfredl** [comment]
An alternative I guess is to call this macro `FOR_ALL_WINDOWS_IN_CURTAB`. (I do not use tabpages very often. I would like that code is explicit if "all windows" actually means "all windows in curtab" or "all windows in all tabs").

**Comment by justinmk** [comment]
This reverts https://github.com/neovim/neovim/pull/1208 , and was discussed elsewhere IIRC. Doesn't seem like a clear win.

**Comment by blueyed** [comment]
> FOR_ALL_WINDOWS_IN_CURTAB

Wouldn't really help to keep the diff smaller - except for when upstreamed, but cannot be done due to (missing) C99 apparently?

> Doesn't seem like a clear win.

Yeah, it's a bit unfortunate, but I still think it is good to have a smaller diff to Vim in general, especially since those macros are used in many places.

With regard to https://github.com/neovim/neovim/pull/1208 I think this should have been proposed to Vim in the first place - I do not think it is worth having a diff in this regard.
What I mean is that it is easier to grasp "FOR_ALL_WINDOWS does visible windows only" instead of having to look up different macros when doing patches etc.

**Comment by justinmk** [comment]
> I still think it is good to have a smaller diff to Vim in general, especially since those macros are used in many places.

In general yes, but not in this case. These macros have never been a problem, even for new contributors.

This is really not worth spending time on.

**Comment by bfredl** [comment]
@blueyed Is is really a good thing to revert clarity improvements just to make merging vim patches marginally simpler? Should we revert the improved names of multibyte functions also? We already have a translation table for renamed names in the wiki. Also "FOR_ALL_WINDOWS does visible windows only" is certainly something  that is subject to change, if we allow "tabgrid" feature in the near future.

**Comment by blueyed** [comment]
> Is is really a good thing to revert clarity improvements just to make merging vim patches marginally simpler?

The benefit for me here is not really visible, and it seemed more like a "drive-by" improvement.
If it is really good for clarity it should have been upstreamed to Vim.

Given how often those macros are used and the subtle differences in them currently (e.g. order of arguments) makes it just a bit more annoying when merging patches / comparing files.

> Also "FOR_ALL_WINDOWS does visible windows only" is certainly something that is subject to change, as we plan to allow "tabgrid" feature in the near future.

Then rephrase this as "does windows in the current tab page".. ;)

What I've meant is that it is similar to e.g. knowing that `win_getid()` etc act on the current tab only (although that one takes a tab by now).

Anyway, I am OK with having this closed still.

@justinmk 
> In general yes, but not in this case. These macros have never been a problem, even for new contributors.

Hard to tell really - I can imagine even new contributors being turned off (without any feedback) due to differences like this, e.g. when trying to help with the patches.

> This is really not worth spending time on.

Yeah.  But it (time) should be considered then maybe in the first place, before creating diff(s) like this (i.e. try to get them in via Vim instead).

As for the wiki: those macros are not mentioned there (https://github.com/neovim/neovim/wiki/Merging-patches-from-upstream-vim#code-differences) - but I also do not think they should (given that it is not really trivial to document, and assuming that this kind of discussion does not turn up again).

**Comment by justinmk** [comment]
> If it is really good for clarity it should have been upstreamed to Vim.

Most things aren't upstreamed, we can't take on that obligation. 

> I can imagine even new contributors being turned off (without any feedback) due to differences like this, e.g. when trying to help with the patches.

The codebases are very different. We merge patches because it's less work than doing the same change starting from zero.

It is definitely good to avoid meaningless differences, but this particular case is small.

**Comment by bfredl** [comment]
@blueyed If it is obvious, as you first say, that ALL_WINDOWS means "windows in the current tab page", then there should be exactly no problem knowing the macros are used for the same thing between the codebases. But to then suddenly say that the difference is so notable that "new contributors are being turned off", then it cannot be clear at all that these constructs mean the same. So how is it then better for contributors to change from a macro that is explicit, to one assumed that new contributors might not understand? If the difference is notable, we should prefer the explicit form; if it is small it doesn't matter either way, also when comparing with vim.

**Comment by blueyed** [comment]
@bfredl 
I've not meant it is obvious, but belongs to the things you need to know / be aware of (and is easy to look up), whereas having to know about the diff is not so obvious (but I am aware of it now at least ;)).

I've meant that contributors are (in my opinion) more likely turned off by having the diff, especially with regard to patches around those macros: typically you would end up looking for why/when it was removed etc.  In my opinion they are less likely turned off by having a macro that is less explicit (given that it is the same in Vim).

> > If it is really good for clarity it should have been upstreamed to Vim.
> 
> Most things aren't upstreamed, we can't take on that obligation.

Sure, but in this case it helps us in the end also.

I can also see that Vim might not be able to use our (shorter) form of the macro(s) (with declarations), and therefore there might be a diff in this regard anyway.
So this is still good/OK for me.. :)

**Comment by bfredl** [comment]
> I've not meant it is obvious, but belongs to the things you need to know / be aware of (and is easy to look up), whereas having to know about the diff is not so obvious (but I am aware of it now at least ;)).

> I've meant that contributors are (in my opinion) more likely turned off by having the diff, especially with regard to patches around those macros: typically you would end up looking for why/when it was removed etc. In my opinion they are less likely turned off by having a macro that is less explicit (given that it is the same in Vim).

I see what you mean, but what you describe as being "turned off" by the difference just means that you have to _learn_ that FOR_ALL_WINDOWS in vim means curtab and not "all windows". So the "thing you need to know" is exactly the same in both cases, we are just making it more explicit so you cannot overlook it by accident. Also, with the introduced float vs popup, the differences with vim here are only going get larger (in nvim they count as part of the window list, unlike vim). Your argument is more relevant for some other part of the codebase where we just maintain vim code as-is (for ui/windows we certainly not).
